ddprof-lib/src/main/cpp/threadFilter.h

```diff
@@ -13,73 +13,80 @@
 #include <vector>
+#include <cstdint>
+#include <memory>
 
-// The size of thread ID bitmap in bytes. Must be at least 64K to allow mmap()
-const u32 BITMAP_SIZE = 65536;
-// How many thread IDs one bitmap can hold
-const u32 BITMAP_CAPACITY = BITMAP_SIZE * 8;
-
-// ThreadFilter query operations must be lock-free and signal-safe;
-// update operations are mostly lock-free, except rare bitmap allocations
 class ThreadFilter {
-private:
-  // Total number of bitmaps required to hold the entire range of thread IDs
-  u32 _max_thread_id;
-  u32 _max_bitmaps;
-  u64 **_bitmap;
-  bool _enabled;
-  volatile int _size;
-
-  u64 *bitmap(int thread_id) {
-    if (thread_id >= _max_thread_id) {
-      return NULL;
-    }
-    return __atomic_load_n(
-        &(_bitmap[static_cast<u32>(thread_id) / BITMAP_CAPACITY]),
-        __ATOMIC_ACQUIRE);
-  }
-
-  static int mapThreadId(int thread_id);
-
-  u64 &word(u64 *bitmap, int thread_id) {
-    // todo: add thread safe APIs
-    return bitmap[((u32)thread_id % BITMAP_CAPACITY) >> 6];
-  }
-
-  u64* const wordAddress(u64 *bitmap, int thread_id) const {
-    return &bitmap[((u32)thread_id % BITMAP_CAPACITY) >> 6];
-  }
-
-  u64* getBitmapFor(int thread_id);
 public:
-  ThreadFilter();
-  ThreadFilter(ThreadFilter &threadFilter) = delete;
-  ~ThreadFilter();
-
-  bool enabled() const { return _enabled; }
-
-  int size() const { return _size; }
-  const volatile int* addressOfSize() const { return &_size; }
-
-  void init(const char *filter);
-  void clear();
+    using SlotID = int;
+
+    // Optimized limits for reasonable memory usage
+    static constexpr int kChunkSize = 256;
+    static constexpr int kChunkShift = 8;   // log2(256)
+    static constexpr int kChunkMask = kChunkSize - 1;
+    static constexpr int kMaxThreads = 2048;
+    static constexpr int kMaxChunks = (kMaxThreads + kChunkSize - 1) / kChunkSize;  // = 8 chunks
+    // High-performance free list using Treiber stack, 64 shards
+    static constexpr int kFreeListSize  = 1024;       // power-of-two for fast modulo
+    static constexpr int kShardCount    = 64;          // power-of-two for fast modulo
+    ThreadFilter();
+    ~ThreadFilter();
+
+    void init(const char* filter);
+    void initFreeList();
+    bool enabled() const;
+    // Hot path methods - slot_id MUST be from registerThread(), undefined behavior otherwise
+    bool accept(SlotID slot_id) const;
+    void add(int tid, SlotID slot_id);
+    void remove(SlotID slot_id);
+    void collect(std::vector<int>& tids) const;
+
+    SlotID registerThread();
+    void unregisterThread(SlotID slot_id);
 
-  inline bool isValid(int thread_id) {
-    return thread_id >= 0 && thread_id < _max_thread_id;
-  }
-
-  bool accept(int thread_id);
-  void add(int thread_id);
-  void remove(int thread_id);
-  u64* bitmapAddressFor(int thread_id);
-
-  void collect(std::vector<int> &v);
+private:
+    // Optimized slot structure with padding to avoid false sharing
+    struct alignas(64) Slot {
```

Comment:
We have definition of `DEFAULT_CACHE_LINE_SIZE` in `dd_arch.h`. I would suggest following code for readability.

```
  struct alignas(DEFAULT_CACHE_LINE_SIZE) Slot {
      std::atomic<int> value{-1};
      char padding[DEFAULT_CACHE_LINE_SIZE - sizeof(value)];
  };
```